modules/determining-upgrade-viability-conditiontype.adoc

```diff
@@ -0,0 +1,43 @@
+// Module included in the following assemblies:
+//
+// * updating/index.adoc
+
+:_content-type: CONCEPT
+[id="Understanding_clusteroperator_conditiontypes_{context}"]
+= Understanding cluster operator condition types
+
+The status of cluster operators is revealed with condition type, informing you of potential impacts to upgrading your cluster. 
+The following condition types are consistent across all ClusterVersion and ClusterOperator conditions. Individual components may also have additional condition types with component-specific language.
+
+The condition type, as well as additional information about your operator, can be retrieved in either YAML or JSON format through the `oc get clusterversion -o` command:
+
+[source,terminal]
+----
+$ oc get clusterversion -o yaml
```

Comment:
`oc get clusterversion -o` might not be a good example because I rarely see degraded condition in clusterversion.

Reply:
Degraded condition is a unhealthy condition , so not seeing it is a good thing. But when it is set then this command is the right command for it. So I disagree with @shellyyang1989

Reply:
okay - thank you both!

Reply:
PTAL https://issues.redhat.com/browse/OCPBUGS-3714 if I recall correctly, CVO would not report Degraded, but would have Failing instead!

Reply:
Right, that's what I meant. CVO reports Failing condition when seeing cluster operators degraded.
Though Degraded=True is not a good thing, the doc is telling what Degraded means, so that we need to tell how Degraded is like by an example. So why not using a common CO yaml which has the four conditions? like authentication

```
# oc get co authentication -ojson | jq -r .status.conditions
[
  {
    "lastTransitionTime": "2023-01-20T02:06:53Z",
    "message": "All is well",
    "reason": "AsExpected",
    "status": "False",
    "type": "Degraded"
  },
  {
    "lastTransitionTime": "2023-01-20T02:14:37Z",
    "message": "AuthenticatorCertKeyProgressing: All is well",
    "reason": "AsExpected",
    "status": "False",
    "type": "Progressing"
  },
  {
    "lastTransitionTime": "2023-01-20T02:06:52Z",
    "message": "All is well",
    "reason": "AsExpected",
    "status": "True",
    "type": "Available"
  },
  {
    "lastTransitionTime": "2023-01-20T01:48:51Z",
    "message": "All is well",
    "reason": "AsExpected",
    "status": "True",
    "type": "Upgradeable"
  }
]
```

Reply:
I also tend to think that the best example would be with all of them visible, unless we have a valid reason for sticking with CVO in this example. 
also, the PR that I've referenced above, mentions:
```
... Degraded is strictly a ClusterOperator thing. ClusterVersion fills a similar role with Failing (although it's not clear to me why folks decided against sticking with the same condition slug for the similar roles).
```
@wking is there a scenario where user can observe Degraded in cvo?  
sorry for the bikeshedding :)

Reply:
May be we can explain the cluster version operator's failing status condition in a followup PR. The failing status condition is little weird as it  only exists for cluster version operator code but does not exists in openshift/api

Reply:
> ... is there a scenario where user can observe Degraded instead of Failing CVO

Nope.  And to back that up for anyone who's skeptical, there's a code search:

```console
$ git remote get-url origin
ssh://github.com/openshift/cluster-version-operator.git
$ git --no-pager log --oneline -1
c5cdf424 (HEAD -> master, origin/release-4.14, origin/release-4.13, origin/master, origin/HEAD) Merge pull request #887 from petr-muller/upgradeable-code-cleanups-1
$ git --no-pager grep Degraded pkg/cvo/status.go
pkg/cvo/status.go:      // cannot reach the desired state. It is considered more serious than Degraded
```

And Telemetry, with:

```
group by (condition) (cluster_operator_conditions{name="version"})
```

giving:

```
{condition="Available"} | 1
{condition="Failing"} | 1
{condition="Progressing"} | 1
{condition="ReleaseAccepted"} | 1
{condition="RetrievedUpdates"} | 1
{condition="Upgradeable"} | 1
{condition="UpgradeableAdminAckRequired"} | 1
{condition="UpgradeableClusterOperators"} | 1
{condition="ImplicitlyEnabledCapabilities"} | 1
{condition="UpgradeableClusterVersionOverrides"} | 1
{condition="Invalid"} | 1
{condition="UpgradeableDeletesInProgress"} | 1
```

Reply:
@evakhoni @shellyyang1989 with this information above, are we clear to move into peer and merge reviews? Any other concerns?

Reply:
@obrown1205 [shellyyang1989](https://github.com/shellyyang1989) is at PTO until next week. from my side this is the only one, however imo it is not solved yet. 
`The following condition types are consistent across all ClusterVersion and ClusterOperator conditions. `
is not the case for Degraded, as mentioned above.
`Condition types, as well as additional information about your operator, can be retrieved in either YAML or JSON format through the oc get clusterversion -o command`
again what exactly is `your operator` in this case? the command mentioned is specific to cluster version operator, (which also isn't the best example anyway). i find it a bit ambiguous when we're talking about cluster operators in general, but mentioning cvo as if explicitly the only place to look for conditions.

Reply:
I agree that the sentence `The following condition types are consistent across all ClusterVersion and ClusterOperator conditions` is not 100% correct

Reply:
Here are  my suggestions for the first paragraph is as follows
```
The status of cluster operators is revealed with condition type, informing you of potential impacts to upgrading your cluster.  We are listing some common ClusterOperator conditions types. Some operators may have additional condition types with operator-specific language which are not explained in this document. Cluster version operator is responsible for collecting these status conditions from cluster operators and represents these is an way which will be easier for OCP administrators to understand the state of the OCP cluster. 

Condition types, as well as additional information about the state of the cluster can be retrieved in either YAML or JSON format through the oc get clusterversion -o command:
```

Reply:
Applied the change - what do we want to do about the clusterversion command @LalatenduMohanty ?

Reply:
![Screenshot from 2023-01-30 11-43-59](https://user-images.githubusercontent.com/3125299/215541918-7e4bc4db-d7f0-4371-a235-cd6013cf0dcd.png)
Here is an example of how the status conditions shows up in webconsole

Reply:
@LalatenduMohanty Just making sure I understand - does this mean we should replace Degraded with Failing? And add the RetrievedUpdates, ReleaseAccepted, and ImplicitlyEnabledCapabilities conditions as well? If these are fairly common, it makes sense to me to add them. 

For this first iteration, what would you prefer to see?

Reply:
I notice that the message says "Cluster operator machine-config is degraded. Are we under the assumption that the reader is looking beyond the message and to the condition itself? I personally suspect that a user would look for the term "Failing" rather than Degraded, if that's the actual condition they receive.

Reply:
> Just making sure I understand - does this mean we should replace Degraded with Failing?

Nope, degraded is an operator condition type. But CVO has an extra type as failing and it communicates this way.  So degraded as a condition type is still valid. As I suggested before we can explain `Failing` condition type later as a followup PR.

Reply:
> For this first iteration, what would you prefer to see?

Lets track that in a separate Jira.  I am fine with explaining any condition type which is not easy to understand from its name.

Reply:
@shellyyang1989 @evakhoni @LalatenduMohanty @wking I've created https://issues.redhat.com/browse/OTA-872 to address the additions to condition types. The team would like to address these in a future PR.

@shellyyang1989 - I'm going to go ahead with the peer review and merge review processes. If you have any other concerns, please let me know and we can address them as well :) Can I consider this PR QE approved at this point with the knowledge we will add more in a future PR?

Reply:
@obrown1205 It looks good to me. Thanks!